### ISSUE TITLE
SSL Cert Error

### ISSUE THREAD

**Issue by regner** [theforeman/puppet-foreman#590]
Using the module to create a new Forman install, and your puppet module to setup puppet as well.

This is my forman server profile:

```
# == Class: profiles::foreman::server
#
# Profile class for configuring the Foreman Server
#
# === Parameters
#
# This is a profile class and is therefore parameter-less. Gets included.
#
class profiles::foreman::server {
  # Hiera lookups
  $config = lookup(foreman::server::config)

  class { '::puppet':
    server              => true,
    autosign_entries    => ['*.example.com'],
    server_environments => [],
    server_reports      => 'foreman',
  }

  class { '::foreman':
    admin_username        => 'example',
    admin_password        => $config[admin_password],
    admin_first_name      => 'Dev',
    admin_last_name       => '',
    admin_email           => 'dev@example.com',
    puppetrun             => true,
    oauth_consumer_key    => $config[oauth_consumer_key],
    oauth_consumer_secret => $config[oauth_consumer_secret],
    configure_epel_repo   => false,
  }
}
```

For the most part the install seems to go ahead and work. However it fail to start Apache. Apache seems to think that the cert is missing.

```
==> foreman01.local1.example.com: Error: Systemd start for httpd failed!
==> foreman01.local1.example.com: journalctl log for httpd:
==> foreman01.local1.example.com: -- Logs begin at Tue 2017-09-19 17:26:14 UTC, end at Tue 2017-09-19 17:39:13 UTC. --
==> foreman01.local1.example.com: Sep 19 17:39:12 foreman01.local1.example.com systemd[1]: Starting The Apache HTTP Server...
==> foreman01.local1.example.com: Sep 19 17:39:12 foreman01.local1.example.com httpd[19716]: AH00526: Syntax error on line 33 of /etc/httpd/conf.d/05-foreman-ssl.conf:
==> foreman01.local1.example.com: Sep 19 17:39:12 foreman01.local1.example.com httpd[19716]: SSLCertificateFile: file '/etc/puppetlabs/puppet/ssl/certs/foreman01.local1.example.com.pem' does not exist or is empty
==> foreman01.local1.example.com: Sep 19 17:39:12 foreman01.local1.example.com systemd[1]: httpd.service: main process exited, code=exited, status=1/FAILURE
==> foreman01.local1.example.com: Sep 19 17:39:12 foreman01.local1.example.com kill[19717]: kill: cannot find process ""
==> foreman01.local1.example.com: Sep 19 17:39:13 foreman01.local1.example.com systemd[1]: httpd.service: control process exited, code=exited status=1
==> foreman01.local1.example.com: Sep 19 17:39:13 foreman01.local1.example.com systemd[1]: Failed to start The Apache HTTP Server.
==> foreman01.local1.example.com: Sep 19 17:39:13 foreman01.local1.example.com systemd[1]: Unit httpd.service entered failed state.
==> foreman01.local1.example.com: Sep 19 17:39:13 foreman01.local1.example.com systemd[1]: httpd.service failed.
```

If I do an `ls` in the location that Apache is looking for the cert however I see it is there:
```
[root@foreman01 ~]# ls -la /etc/puppetlabs/puppet/ssl/certs/
total 8
drwxr-xr-x. 2 puppet puppet   59 Sep 19 17:30 .
drwxrwx--x. 8 puppet puppet  126 Sep 19 17:30 ..
-rw-r--r--. 1 puppet puppet 2000 Sep 19 17:30 ca.pem
-rw-r--r--. 1 puppet puppet 2045 Sep 19 17:30 foreman01.local1.example.com.pem
```

What am I missing?

**Comment by ekohl** [comment]
My suspicion would be that they are generated after Apache is started. You can try to let foreman require puppet.

**Comment by Regner** [comment]
Thanks, will give that a shot now. :)

**Comment by Regner** [comment]
Here is how I attempted to do as you recommended. I moved the Puppet configuration into a separate manifest and then in the role imported it and did the ordering there.

```
# == Class: roles::foreman::server
#
# This is a role class to configure a Foreman server.
#
class roles::foreman::server {

  include ::profiles::common
  include ::profiles::foreman::puppet
  include ::profiles::foreman::server

  Class['::profiles::common'] ->
  Class['::profiles::foreman::puppet'] ->
  Class['::profiles::foreman::server']
}
```

I am still getting the same error however.

**Comment by ekohl** [comment]
I assumed you used puppet-puppet, but can you confirm this? Another thing I wonder is if you get any other errors and if you can whether see ```Exec['puppet_server_config-generate_ca_cert']``` happens before or after the Apache service.

A last thing to check is whether the resources are actually chained. If you use ```include``` rather than ```contain``` the chaining might not have an effect.

**Comment by Regner** [comment]
Using `theforeman/puppet`.

It looks like that exec is happening before the apache service stuff. Here are the logs from the "Applying configuration version..." up to the exec command you listed though:
```
==> foreman01.local1.example.com: Info: Applying configuration version '1505853920'
==> foreman01.local1.example.com: Debug: /Stage[main]/Profiles::Common/before: subscribes to Class[Profiles::Foreman::Puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel-source]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel-debuginfo]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel-testing]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel-testing-source]
==> foreman01.local1.example.com: Debug: /Stage[main]/Epel/Epel::Rpm_gpg_key[EPEL-7]/before: subscribes to Yumrepo[epel-testing-debuginfo]
==> foreman01.local1.example.com: Debug: /Stage[main]/Profiles::Foreman::Puppet/before: subscribes to Class[Profiles::Foreman::Server]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Config/before: subscribes to Class[Puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Config/notify: subscribes to Class[Puppet::Agent::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Config/notify: subscribes to Class[Puppet::Server::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Config/File[/etc/puppetlabs/puppet]/before: subscribes to Concat[/etc/puppetlabs/puppet/puppet.conf]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Config/Concat[/etc/puppetlabs/puppet/puppet.conf]/notify: subscribes to File[/etc/puppetlabs/puppet/auth.conf]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent/before: subscribes to Class[Puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Install/before: subscribes to Class[Puppet::Agent::Config]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Config/notify: subscribes to Class[Puppet::Agent::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service/before: subscribes to Class[Puppet::Agent]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service/Anchor[puppet::agent::service_start]/before: subscribes to Class[Puppet::Agent::Service::Daemon]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service/Anchor[puppet::agent::service_start]/before: subscribes to Class[Puppet::Agent::Service::Systemd]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service/Anchor[puppet::agent::service_start]/before: subscribes to Class[Puppet::Agent::Service::Cron]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Daemon/before: subscribes to Anchor[puppet::agent::service_end]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Systemd/before: subscribes to Anchor[puppet::agent::service_end]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Systemd/Service[puppet-run.timer]/before: subscribes to File[/etc/systemd/system/puppet-run.timer]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Systemd/Service[puppet-run.timer]/before: subscribes to File[/etc/systemd/system/puppet-run.service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Systemd/File[/etc/systemd/system/puppet-run.timer]/notify: subscribes to Exec[systemctl-daemon-reload-puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Systemd/File[/etc/systemd/system/puppet-run.service]/notify: subscribes to Exec[systemctl-daemon-reload-puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Agent::Service::Cron/before: subscribes to Anchor[puppet::agent::service_end]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server/before: subscribes to Class[Puppet]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server/before: subscribes to Class[Puppet::Agent::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Install/before: subscribes to Class[Foreman::Config]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Install/notify: subscribes to Class[Puppet::Server::Config]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Config/before: subscribes to Class[Foreman::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Config/notify: subscribes to Class[Puppet::Server::Service]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Config/File[/etc/puppetlabs/puppet/ssl/private_keys]/require: subscribes to Exec[puppet_server_config-create_ssl_dir]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/require: subscribes to Concat[/etc/puppetlabs/puppet/puppet.conf]
==> foreman01.local1.example.com: Debug: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/require: subscribes to Exec[puppet_server_config-create_ssl_dir]
```

I am still relatively new to Puppet so not entirely sure I am doing things right for the chaining.

I shared above the role, here is the puppet and foreman profiles that it includes:
```
class profiles::foreman::puppet {
  class { '::puppet':
    server              => true,
    autosign_entries    => ['*.example.com'],
    server_environments => [],
    server_reports      => 'foreman',
  }
}
```

```
class profiles::foreman::server {
  # Hiera lookups
  $config = lookup(foreman::server::config)

  class { '::foreman':
    admin_username        => 'voxter',
    admin_password        => $config[admin_password],
    admin_first_name      => 'Dev',
    admin_last_name       => '',
    admin_email           => 'dev@example.com',
    puppetrun             => true,
    oauth_consumer_key    => $config[oauth_consumer_key],
    oauth_consumer_secret => $config[oauth_consumer_secret],
    configure_epel_repo   => false,
  }
}
```

Thanks once again for all your help with this. It is much appreciated.

**Comment by ekohl** [comment]
It could be we don't properly contain resources. Just to verify it can you explicitly chain the ```Exec``` to ```Service['httpd']``` and see if it works? When that does work, I'd consider it a bug somewhere in our code.

**Comment by Regner** [comment]
Any chance you could give me an example of how to do that?

**Comment by ekohl** [comment]
I meant an explicit ```Exec['puppet_server_config-generate_ca_cert'] -> Service['httpd']```

**Comment by Regner** [comment]
Got the same error.

I don't know how I missed it before, but I also seem to be getting this error: 
```
Notice: /Stage[main]/Puppet::Server::Service/Service[puppetserver]/enable: enable changed 'false' to 'true'
Error: /Stage[main]/Puppet::Server::Service/Service[puppetserver]: Failed to call refresh: Systemd restart for puppetserver failed!
journalctl log for puppetserver:
-- Logs begin at Wed 2017-09-20 23:05:11 UTC, end at Thu 2017-09-21 04:40:00 UTC. --
Sep 21 04:39:35 foreman01.local1.example.com systemd[1]: Stopping puppetserver Service...
Sep 21 04:39:36 foreman01.local1.example.com systemd[1]: Starting puppetserver Service...
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: Exception in thread "main" java.lang.RuntimeException: Service ':FilesystemWatchService' not found
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$handle_prismatic_exception_BANG_.invokeStatic(internal.clj:121)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$handle_prismatic_exception_BANG_.invoke(internal.clj:111)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$instantiate$fn__13800.invoke(internal.clj:162)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$instantiate.invokeStatic(internal.clj:159)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$instantiate.invoke(internal.clj:152)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$eval14334$build_app_STAR___14343$fn__14344.invoke(internal.clj:569)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$eval14334$build_app_STAR___14343.invoke(internal.clj:537)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445$fn__14446$fn__14447.invoke(internal.clj:639)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445$fn__14446.invoke(internal.clj:638)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.internal$eval14436$boot_services_STAR___14445.invoke(internal.clj:633)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$eval15266$boot_with_cli_data__15273$fn__15274.invoke(core.clj:130)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$eval15266$boot_with_cli_data__15273.invoke(core.clj:95)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$eval15295$run__15300$fn__15301.invoke(core.clj:151)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$eval15295$run__15300.invoke(core.clj:145)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$main.invokeStatic(core.clj:173)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.core$main.doInvoke(core.clj:159)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.RestFn.invoke(RestFn.java:512)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.invoke(Var.java:409)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.AFn.applyToHelper(AFn.java:178)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.applyTo(Var.java:700)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.core$apply.invokeStatic(core.clj:646)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.core$apply.invoke(core.clj:641)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.main$_main.invokeStatic(main.clj:7)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at puppetlabs.trapperkeeper.main$_main.doInvoke(main.clj:4)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.RestFn.invoke(RestFn.java:512)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.invoke(Var.java:409)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.AFn.applyToHelper(AFn.java:178)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.applyTo(Var.java:700)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.core$apply.invokeStatic(core.clj:646)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.main$main_opt.invokeStatic(main.clj:314)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.main$main_opt.invoke(main.clj:310)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.main$main.invokeStatic(main.clj:421)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.main$main.doInvoke(main.clj:384)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.RestFn.invoke(RestFn.java:619)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.invoke(Var.java:430)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.AFn.applyToHelper(AFn.java:195)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.lang.Var.applyTo(Var.java:700)
Sep 21 04:40:00 foreman01.local1.example.com puppetserver[19397]: at clojure.main.main(main.java:37)
Sep 21 04:40:00 foreman01.local1.example.com systemd[1]: puppetserver.service: control process exited, code=exited status=1
Sep 21 04:40:00 foreman01.local1.example.com systemd[1]: Failed to start puppetserver Service.
Sep 21 04:40:00 foreman01.local1.example.com systemd[1]: Unit puppetserver.service entered failed state.
Sep 21 04:40:00 foreman01.local1.example.com systemd[1]: puppetserver.service failed.
```

**Comment by ekohl** [comment]
That sounds like you're using puppetserver 5.1 with puppet-puppet < 8.0.3

**Comment by Regner** [comment]
Bumping the version of `theforeman/puppet` to `8.0.3` fixed that filewatch service but I still get the cert issue.

Is there any chance this is a permissions issue? With the cert file being owned by puppet? It looks like read permissions are set correctly though...

**Comment by Regner** [comment]
Not sure how to solve it, but the issue relates to SELinux disallowing the Apache user from accessing the file.

**Comment by Regner** [comment]
Running `setenforce 0` before starting the puppet agent run fixed the issue. Have run into another issue, but that seems to be the way to solve this issue. Not sure what needs to happen in the puppet module to make that work though.

**Comment by ekohl** [comment]
That's a good one I hadn't thought about. I'll see if I can replicate it.

**Comment by Regner** [comment]
Let me know if you have issues reproducing it. I can share a Vagrant file and Puppet config that will cause it.

**Comment by wsommers** [comment]
has anyone come up with another solution besides disabling selinux enforcement?

**Comment by ekohl** [comment]
Quite a lot has changed inside our modules. I'm closing this now due to its age.
